File: DiseaseMutationsApp/SequenceRepository.cs
namespace DiseaseMutationsApp;

public class SequenceRepository
{
    private const string BASE_URL = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&id=";

    private static readonly Dictionary<string, Sequence> sequences = new();
    
    public static async Task<Sequence> GetSequence(string id)
    {
        if (sequences.TryGetValue(id, out var sequence))
        {
            return sequence;
        }

        sequence = new Sequence(id, await GetSequenceData(id));
        sequences[id] = sequence;
        return sequence;
    }

    private static async Task<string> GetSequenceData(string id)
    {
        var httpClient = new HttpClient();
        var response = await httpClient.GetAsync($"{BASE_URL}{id}&rettype=fasta");
        response.EnsureSuccessStatusCode();
        var content = await response.Content.ReadAsStringAsync();
        var lines = content.Split('\n');
        var data = string.Join("", lines.Skip(1)).Trim();
        return data;
    }
}

File: gRNA/SequenceRepository.fs
module gRNA.SequenceRepository

open System.Collections.Generic
open System.Net.Http
open System.Threading.Tasks
open gRNA.Sequence

type SequenceRepository() =
    static let BASE_URL = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&id="
    static let sequences = Dictionary<string, Sequence>()
    
    static member GetSequence(id: string) =
        task {
            match sequences.TryGetValue(id) with
            | true, sequence -> 
                return sequence
            | false, _ ->
                let! data : string = SequenceRepository.GetSequenceData(id)
                let rna = data.Replace('T', 'U')
                let sequence = Sequence(id, rna)
                sequences.[id] <- sequence
                return sequence
        }
    
    static member private GetSequenceData(id: string) =
        task {
            use httpClient = new HttpClient()
            let! response = httpClient.GetAsync($"{BASE_URL}{id}&rettype=fasta")
            response.EnsureSuccessStatusCode() |> ignore
            let! content = response.Content.ReadAsStringAsync()
            let lines = content.Split('\n')
            let data =
                lines 
                |> Array.skip 1 
                |> String.concat ""
                |> _.Trim()
            return data
        }